### ISSUE TITLE
J2SE instrumentation 

### ISSUE THREAD

**Issue by madushan1000** [frida/frida-java-bridge#72]
I have been looking for java instrumentation frameworks and it has been a tedious journey to nowhere. 
I was wondering if we can use frida to instrument java code running on x86 processors(windows/linux) without the android/dalvik runtime. If we can't, what's stopping us from doing that?

**Comment by ChaosData** [comment]
I don't know of a good way to hook (non-Android JVM) Java functions from Frida, but one can invoke the [JNI APIs](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html) to essentially perform reflection. For example (this just gets you to the point that you can wrap and invoke arbitrary JNI APIs by their offset in the list I linked; JNI-based reflection is a pain, even in C):

```JavaScript
  var JNI_GetCreatedJavaVMs;

  Module.enumerateExports("libjvm.so", {
    "onMatch": function(ex) {
      if (ex.name == "JNI_GetCreatedJavaVMs") {
        JNI_GetCreatedJavaVMs = new NativeFunction(
          ex.address, 'int', ['pointer', 'pointer', 'pointer']
        );
        return "stop";
      }
    },
    "onComplete": function() {}
  });

  var vm_buf = Memory.alloc(Process.pointerSize);
  var bufLen = ptr(1);
  var nVMs = Memory.alloc(Process.pointerSize);

  if (JNI_GetCreatedJavaVMs(vm_buf, bufLen, nVMs) !== 0) {
    return;
  }

  if (parseInt(Memory.readPointer(nVMs)) !== 1) {
    return;
  }

  var JavaVM = Memory.readPointer(vm_buf);
  var JavaVM_functions = Memory.readPointer(JavaVM);

  var AttachCurrentThread = new NativeFunction(
    Memory.readPointer(JavaVM_functions.add(4*Process.pointerSize)),
    'int',
    ['pointer', 'pointer', 'pointer']
  );

  var env_buf = Memory.alloc(Process.pointerSize);

  if (AttachCurrentThread(JavaVM, env_buf, NULL) !== 0) {
    return;
  }

  var JNIEnv = Memory.readPointer(env_buf);
  var JNIEnv_functions = Memory.readPointer(JNIEnv);

  var GetVersion = new NativeFunction(
    Memory.readPointer(JNIEnv_functions.add(4*Process.pointerSize)),
    'int',
    ['pointer']
  );

  console.log(GetVersion(JNIEnv));
```

Other than that though, you probably want to look into using Java agents for bytecode instrumentation. [raphw/byte-buddy](https://github.com/raphw/byte-buddy) is a solid choice for building such agents. For the most part, you can pass the path to an agent JAR configured with the special manifest attributes in a `-javaagent` argument that will cause it to run before `main` has. You can then modify classes as they load to inject arbitrary functionality.

**Comment by madushan1000** [comment]
I'm wondering if we can't just enable the android specific JAVA api which are already in Frida for x86 too. I'm guessing there are some non portable code in the hooking method, etc.. (or worse if the JAVA disassembly engine only works for dalvik/ART).
I was aware of the javaagent method but thought I'd first check out Frida since I have experience with it. I'm going to try javaagent as a last resort. Thank you for the suggestion.

**Comment by ChaosData** [comment]
The current implementation of hooking is tied to android/art, though the concept of making functions native to reroute them to Frida JS hooks probably would work on any JVM, though I'm not sure JVMTI provides and out-of-the-box way to make Java methods native. There would need to be an equivalent for whatever JVM was being targeted (generally HotSpot). The class wrapper handling in class-factory.js is also fairly android/dalvik specific. I'm just musing at this point though.

**Comment by oleavr** [comment]
@ChaosData It shouldn't be hard to improve `frida-java` so it's able to provide all of the introspection on other VMs – method hooking could be the second milestone. Most of its code is just making JNI calls like in your example, and the goal is to support non-Android VMs too. (It's why I renamed the global from `Dalvik` to `Java` some years back.)

**Comment by jhftss** [comment]
@oleavr When will it support the JVM?

**Comment by oleavr** [comment]
@jhftss When somebody contributes support for it.

**Comment by nevinhappy** [comment]
So this issue is still open.

**Comment by tr33s** [comment]
Has this been implemented with 12.10?

**Comment by oleavr** [comment]
Yes, we can finally close this issue! 🎉 More details here:
https://frida.re/news/2020/06/29/frida-12-10-released/
